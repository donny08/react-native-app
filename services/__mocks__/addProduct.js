const fakeResponse = {
  success: true,
  response: 'Applied for product successfully!'
};

export default async payload => {
    return await new Promise(resolve => {
      resolve(fakeResponse);
    });
};