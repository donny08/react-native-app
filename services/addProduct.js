const authorization ='Basic eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJkZW1vdXNlcjEiLCJuYW1lIjoiUkFLQkFOSyBEZW1vIiwiaWF0IjoxNTE2MjM5MDIyfQ.hD2So1Jms00c7XB_bxNIrTgvSqMesmrlc0FkBAdY6rM';

export default async payload => {
  try {
    let response = await fetch('http://rakbank-test.mocklab.io/activation', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: authorization,
      },
      body: JSON.stringify(payload),
    });
    let responseJson = await response.json();
    return {
      success: true,
      response: responseJson.message,
    };
  } catch (error) {
    console.error(error);
    return {
      success: false,
      error: error,
    };
  }
};
// import 'isomorphic-fetch';

// class HomeService {
//   constructor() {
//     this.authorization =
//       'Basic eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJkZW1vdXNlcjEiLCJuYW1lIjoiUkFLQkFOSyBEZW1vIiwiaWF0IjoxNTE2MjM5MDIyfQ.hD2So1Jms00c7XB_bxNIrTgvSqMesmrlc0FkBAdY6rM';
//   }

//   async addProduct(payload) {
//     try {
//       let response = await fetch('http://rakbank-test.mocklab.io/activation', {
//         method: 'POST',
//         headers: {
//           Accept: 'application/json',
//           'Content-Type': 'application/json',
//           Authorization: this.authorization,
//         },
//         body: JSON.stringify(payload),
//       });
//       let responseJson = await response.json();
//       return {
//         success: true,
//         response: responseJson.message,
//       };
//     } catch (error) {
//       console.error(error);
//       return {
//         success: false,
//         error: error,
//       };
//     }
//   }
// }
// const homeService = new HomeService();

// export default homeService;
