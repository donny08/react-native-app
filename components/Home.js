import React, {Component} from 'react';
import {
  StyleSheet,
  Text,
  View,
  // ScrollView,
  // FlatList,
  Image,
  TouchableWithoutFeedback,
  ActivityIndicator,
  // Button,
  Easing,
  Animated,
  SafeAreaView,
  Alert,
} from 'react-native';
import addProduct from '../services/addProduct';
import MaterialIndicator from './MaterialIndicator';
import STATUS from '../services/status';

// import {SafeAreaView, SafeAreaProvider } from 'react-native-safe-area-view';

class Home extends Component {
  state = {
    isLoading: true,
    title: STATUS.ACTIVATE,
    image: require('../images/arrow.png'),
    color: 'rgb(0,0,255)',
    icon: {
      width: 16,
      height: 16,
      marginRight: 10,
    },
  };

  constructor(props) {
    super(props);
    this.handlePressIn = this.handlePressIn.bind(this);
    this.handlePressOut = this.handlePressOut.bind(this);
    this.makeAPIRequest = this.makeAPIRequest.bind(this);
    this.animatedValue = new Animated.Value(1);
    this.animatedText = new Animated.Value(0);
  }
  componentDidMount() {
    // console.log(Status);
  }

  async makeAPIRequest() {
    try {
      const product = await addProduct({
        productId: '82jqp008d2l00',
        emirate: 'Abu Dhabi',
      });

      if (product && product.success) {
        Animated.timing(this.animatedText, {
          toValue: -100,
          easing: Easing.back(1),
          duration: 1000,
        }).start(() => {
          this.animatedText.setValue(0);
          this.setState({
            title: STATUS.ACTIVATED,
            image: require('../images/tick.png'),
            color: '#51da7c',
            icon: {
              width: 18,
              height: 18,
              marginRight: 5,
              marginTop: -1,
            },
          });
        });
      }
    } catch (error) {
      Alert.alert('Oops! Something went wrong. Please try again.');
      this.setState({
        title: STATUS.ACTIVATE,
        image: require('../images/arrow.png'),
        color: 'rgb(0,0,255)',
        icon: {
          width: 16,
          height: 16,
          marginRight: 10,
        },
      });
    }
  }

  handlePressIn() {
    // this.animatedValue.setValue(0);
    Animated.spring(this.animatedValue, {
      toValue: 0.7,
    }).start();
    // Animated.timing(this.animatedValue, {
    //   toValue: 0.9,
    // }).start();

    if (this.state.title == 'Activate') {
      Animated.timing(this.animatedText, {
        toValue: -100,
        easing: Easing.back(1),
        duration: 1000,
      }).start(() => {
        this.animatedText.setValue(0);
      });

      this.setState({
        title: STATUS.WAITING,
        image: require('../images/spinner.gif'),
        color: 'rgb(0,0,255)',
        icon: {
          width: 16,
          height: 16,
          marginRight: 8,
        },
      });
      this.makeAPIRequest();
    }
  }

  handlePressOut() {
    Animated.spring(this.animatedValue, {
      toValue: 1,
      friction: 3,
      tension: 40,
    }).start();
  }

  render() {
    const {title, image, color, icon} = this.state;
    const animatedStyle = {
      transform: [{scale: this.animatedValue}],
    };

    const animatedText = {
      transform: [
        {
          translateY: this.animatedText,
        },
      ],
    };
    // const animatedText = {
    //   transform: [
    //     {
    //       translateY: this.animatedValue.interpolate({
    //         inputRange: [0, 1],
    //         outputRange: [-150, 1],
    //         extrapolate: 'clamp',
    //       }),
    //     },
    //   ],
    // };
    return (
      <View style={styles.container}>
        <Image style={styles.logo} source={require('../images/logo.png')} />
        <TouchableWithoutFeedback
          onPressIn={this.handlePressIn}
          onPressOut={this.handlePressOut}>
          <Animated.View
            style={[styles.button, animatedStyle, {backgroundColor: color}]}>
            {/* {title != STATUS.WAITING && <Image style={icon} source={image} />}
            {title == STATUS.WAITING && (
              <MaterialIndicator color="white" size={16} />
            )} */}
            <Animated.View style={[styles.row, animatedText]}>
              {title != STATUS.WAITING && <Image style={icon} source={image} />}
              {title == STATUS.WAITING && (
                <MaterialIndicator color="white" size={16} />
              )}
              <Text style={styles.text}>{title}</Text>
            </Animated.View>

            {/* <Animated.Text style={[styles.text, animatedText]}>
              {title}
            </Animated.Text> */}
          </Animated.View>
        </TouchableWithoutFeedback>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  logo: {
    width: 100,
    height: 100,
    borderRadius: 20,
    marginBottom: 100,
  },
  button: {
    width: 100,
    height: 35,
    borderRadius: 20,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  row: {
    flexDirection: 'row',
  },
  text: {
    color: '#FFF',
    fontWeight: '700',
    fontSize: 12,
  },
});

export default Home;
