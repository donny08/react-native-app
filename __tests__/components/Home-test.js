// import 'react-native';
import React from 'react';
import {shallow, mount} from 'enzyme';
import Home from '../../components/Home';
import MaterialIndicator from '../../components/MaterialIndicator';
import STATUS from '../../services/status';
import TestRenderer from 'react-test-renderer';
import {TouchableWithoutFeedback, Animated, Text} from 'react-native';
// import jsdom from 'jsdom';
// require('react-native-mock-render/mock');

// var jsdom = require('jsdom');
// const {JSDOM} = jsdom;

// const {document} = new JSDOM('').window;
// global.document = document;
// global.window = document.defaultView;
// Object.keys(document.defaultView).forEach(property => {
//   if (typeof global[property] === 'undefined') {
//     global[property] = document.defaultView[property];
//   }
// }); 
jest.mock('../../services/addProduct');

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import addProduct from '../../services/addProduct';

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

describe('Home Component Test', () => {
  let wrapper, testRenderer, testInstance;

   beforeEach(() => {
     //homeService.mockClear();
    fetch.resetMocks();
    wrapper = shallow(<Home />);
    testRenderer = TestRenderer.create(<Home />);
    testInstance = testRenderer.root;
   });

   afterEach(() => {
     jest.resetAllMocks();
   });

  it('renders correctly', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('display text "Active" on button', () => {
    const button = testInstance.findAllByType(Text)[0];
    // console.log(button.props.children);
    expect(button.props.children).toEqual(STATUS.ACTIVATE);
  });

  // it('tap on <TouchableWithoutFeedback /> and expect return <MaterialIndicator />', () => {
  //   const wrapper = shallow(<Home />);
  //   wrapper
  //     .find(TouchableWithoutFeedback)
  //     .first()
  //     .props()
  //     .onPressIn();
      
  //   expect(wrapper.find(MaterialIndicator).length).toBe(1);
  // });

  it('tap on <TouchableWithoutFeedback /> and make API call', async () => {
    const button = testInstance
      .findByType(TouchableWithoutFeedback)
      .props
      .onPressIn;
    console.debug(button);
    expect(button).toHaveBeenCalledTimes(1);
      //  addProduct.mockReturnValue({
      //    json: () => 'it worked!',
      //  });

        // const response = await Home.makeAPIRequest()
        // expect(response).toEqual('it worked!');
  });
  /* it('tap on <TouchableWithoutFeedback /> and hit API call', async () => {
    const wrapper = shallow(<Home />);
    const mock = jest.fn();

  
    const response = {
      success: true,
      response: 'Applied for product successfully!',
    };

    // const payload = {
    //   productId: '82jqp008d2l00',
    //   emirate: 'Abu Dhabi',
    // };

    mock.mockClear();
    mock.mockReturnValueOnce(response);
    // addProduct(payload).mockClear();
    // addProduct(payload).mockReturnValueOnce(Promise.resolve(response));
    // fetch.mockClear();
    // fetch.mockReturnValueOnce(Promise.resolve(response));

    await wrapper
      .find(TouchableWithoutFeedback)
      .first()
      .props()
      .onPressIn();

    expect(mock).toHaveBeenCalledTimes(1);
    // expect(fetch).toHaveBeenCalledTimes(1);

    await Promise.resolve(response); // added it
    // expect(wrapper.find(MaterialIndicator).length).toBe(1);
    // expect(onPressEvent.mock.calls.length).toBe(1);
    // const responseJson = {
    //   message: 'Applied for product successfully!',
    // };
    // const response = {
    //   success: true,
    //   response: responseJson,
    // };
    // await homeService.addProduct.mockImplementation(() => {
    //   return Promise.resolve(response);
    // });
    // // homeService.addProduct.mockClear();
    // // homeService.addProduct.mockReturnValueOnce(Promise.resolve(response));
    // fetch.mockClear();
    // fetch.mockReturnValueOnce(Promise.resolve(responseJson));

    // wrapper.find(TouchableWithoutFeedback).simulate('press');

    // await flushPromises();

    // expect(homeService.addProduct).toHaveBeenCalledTimes(1);
    // expect(fetch).toHaveBeenCalledTimes(1);
    // expect(wrapper.find(MaterialIndicator).length).toBe(1);

    // await Promise.resolve(); // added it
    // expect(onUpload).toHaveBeenCalledTimes(1);
  });*/
});
