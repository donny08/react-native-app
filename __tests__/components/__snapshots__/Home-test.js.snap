// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Test renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../images/logo.png",
      }
    }
    style={
      Object {
        "borderRadius": 20,
        "height": 100,
        "marginBottom": 100,
        "width": 100,
      }
    }
  />
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(0,0,255)",
        "borderRadius": 20,
        "flexDirection": "row",
        "height": 35,
        "justifyContent": "center",
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "width": 100,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../images/arrow.png",
        }
      }
      style={
        Object {
          "height": 16,
          "marginRight": 0,
          "width": 16,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontSize": 12,
          "fontWeight": "700",
          "transform": Array [
            Object {
              "translateY": 1,
            },
          ],
        }
      }
    >
      Activate
    </Text>
  </View>
</View>
`;
